artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/QueueImpl.java

```diff
@@ -2018,7 +2018,20 @@ public void incrementMesssagesAdded() {
 
    @Override
    public void deliverScheduledMessages() throws ActiveMQException {
-      List<MessageReference> scheduledMessages = scheduledDeliveryHandler.cancel(null);
+      internalDeliverScheduleMessages(scheduledDeliveryHandler.cancel(ref -> true));
+   }
+
+   @Override
+   public void deliverScheduledMessages(String filterString) throws ActiveMQException {
+      internalDeliverScheduleMessages(scheduledDeliveryHandler.cancel(ref -> filterString == null || filterString.length() == 0 ? true : FilterImpl.createFilter(filterString).match(ref.getMessage())));
```

Comment:
The performance could be improved creating the filter only one time before creating the comparator
```suggestion
      final Filter filter = filterString == null || filterString.length() == 0 ? null : FilterImpl.createFilter(filterString);
      internalDeliverScheduleMessages(scheduledDeliveryHandler.cancel(ref -> filter == null ? true : filter.match(ref.getMessage())));
```

Reply:
Great suggestion!